bot/cogs/help_channels.py

```diff
@@ -0,0 +1,665 @@
+            name = self.name_queue.popleft()
+        except IndexError:
+            log.debug("No more names available for new dormant channels.")
+            return None
+
+        log.debug(f"Creating a new dormant channel named {name}.")
+        return await self.dormant_category.create_text_channel(name)
+
+    def create_name_queue(self) -> deque:
+        """Return a queue of element names to use for creating new channels."""
+        log.trace("Creating the chemical element name queue.")
+
+        used_names = self.get_used_names()
+
+        log.trace("Determining the available names.")
+        available_names = (name for name in self.name_positions if name not in used_names)
+
+        log.trace("Populating the name queue with names.")
+        return deque(available_names)
+
+    @commands.command(name="dormant", enabled=False)
+    @with_role(*constants.HelpChannels.cmd_whitelist)
+    async def dormant_command(self, ctx: commands.Context) -> None:
+        """Make the current in-use help channel dormant."""
+        log.trace("dormant command invoked; checking if the channel is in-use.")
+
+        if ctx.channel.category == self.in_use_category:
+            self.cancel_task(ctx.channel.id)
+            await self.move_to_dormant(ctx.channel)
+        else:
+            log.debug(f"{ctx.author} invoked command 'dormant' outside an in-use help channel")
+
+    async def get_available_candidate(self) -> discord.TextChannel:
+        """
+        Return a dormant channel to turn into an available channel.
+
+        If no channel is available, wait indefinitely until one becomes available.
+        """
+        log.trace("Getting an available channel candidate.")
+
+        try:
+            channel = self.channel_queue.get_nowait()
+        except asyncio.QueueEmpty:
+            log.info("No candidate channels in the queue; creating a new channel.")
+            channel = await self.create_dormant()
+
+            if not channel:
+                log.info("Couldn't create a candidate channel; waiting to get one from the queue.")
+                await self.notify()
+                channel = await self.wait_for_dormant_channel()
+
+        return channel
+
+    def get_alphabetical_position(self, channel: discord.TextChannel) -> t.Optional[int]:
+        """
+        Return the position to move `channel` to so alphabetic order is maintained.
+
+        If the channel does not have a valid name with a chemical element, return None.
+        """
+        log.trace(f"Getting alphabetical position for #{channel} ({channel.id}).")
+
+        try:
+            position = self.name_positions[channel.name]
+        except KeyError:
+            log.warning(f"Channel #{channel} ({channel.id}) doesn't have a valid name.")
+            position = None
+
+        log.trace(
+            f"Position of #{channel} ({channel.id}) in Dormant will be {position} "
+            f"(was {channel.position})."
+        )
+
+        return position
```

Comment:
As we've discovered, this approach does not work. After doing a bit of testing, I found out it's because Discord channels the position integers of channels that are already in the category: If you add a channel to a category, the position of all channels with a higher (or equal?) position integer than the one you're inserting will be increased by one:

```
Moving channel h to position 8
Before move: [('b', 2), ('e', 5), ('f', 7),           ('i', 10), ('g', 11), ('j', 13), ('m', 19)]
After move:  [('b', 2), ('e', 5), ('f', 7), ('h', 8), ('i', 11), ('g', 12), ('j', 14), ('m', 20)]
```

As you can see, all the channels with a position `int` higher than 8 had their position integer increased by one. I can't guarantee that this will always happen, but after quite a few experiments, I've never observed an instance that violated this pattern.

My first approach to circumvent this was by assigning each channel a very large starting int (multiples of 200) to give it a buffer against going over the position int of the next channel. However, since Discord minimizes the integers down to the current largest used integer plus one, this doesn't work at all.

Another approach does seem to work: If you insert a channel at the position int of a channel that's already there, it seems to be that it will reliably be inserted before the existing channel. This means that if you have a channel called "c" with position integer 2 and you want to insert "b", you can insert "b" at 2 and the sorting order appears to end up just like we want it. Again, I can only go by experimentation, but it does seem to work.

I used something like this to do the work:
```py
import bisect

def get_position(channel: discord.TextChannel, destination: discord.CategoryChannel) -> int:
    # If the destination category is empty, return the first position
    if  not destination.channels:
        return 1

    # Make list of channel names for bisect
    channel_names = [c.name for c in destination.channels]

    # Get the bisect location of the channel
    rank = bisect.bisect(channel_names, channel.name)

    # If rank is equal to the length, we should insert the channel at the end, otherwise
    # we insert it at the location of the channel it should be inserted before.
    if rank == len(destination.channels):
        return destination.channels[-1].position + 1
    else:
        return destination.channels[rank].position
```

If a channel ever falls out of order, this procedure won't correct that, although it will get back into order once the "out-of-order" channels get cycled through the available/active categories.

Again, it may be that this system is not entirely robust either, I can only go by experimentation, but it seems to work as expected in the tests I've done.

Reply:
I really appreciate you looking into this. I'll try to implement your code into the cog.

Reply:
It works! But what would happen if the `channel_names` list is unsorted or partially sorted? discord.py sorts the category's channels by position rather than by name. Would it be better to sort them by name ourselves?

Reply:
If the `channel_names` list is partially sorted and/or entirely out of order, we will slowly start building an order and it will get back into order at the latest when every channel has been cycled once. I'm not quite sure what sorting the list manually would do, except that it will guarantee that we then insert a channel before the one it's supposed to come for.

That could be preferable, but it won't be a magic "sorting fix". We could also try to detect if the thing is out of order and, if so, try to fix it with the minimum number of channel edits as is possible, but I'd have to have to sit down and thing of a clever way of doing that. There's probably an algorithm for that.

Reply:
I'm fine with it as it is since it will eventually fix itself. It should only get out of order if someone accidentally moves a channel (or if the implementation turns out to not always work). An separate issue can be opened to tackle this in a more clever way.